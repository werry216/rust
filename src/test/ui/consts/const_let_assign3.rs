#![feature(const_fn)]

struct S {
    state: u32,
}

impl S {
    const fn foo(&mut self, x: u32) {
        self.state = x;
        //~^ contains unimplemented expression
    }
}

const FOO: S = {
    let mut s = S { state: 42 };
    s.foo(3); //~ ERROR references in constants may only refer to immutable values
    s
};

type Array = [u32; { let mut x = 2; let y = &mut x; *y = 42; *y}];
//~^ ERROR references in constants may only refer to immutable values
//~| ERROR constant contains unimplemented expression type

fn main() {
    assert_eq!(FOO.state, 3);
}
