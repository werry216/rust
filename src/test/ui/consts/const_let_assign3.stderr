error[E0019]: constant function contains unimplemented expression type
  --> $DIR/const_let_assign3.rs:9:9
   |
LL |         self.state = x;
   |         ^^^^^^^^^^^^^^

error[E0017]: references in constants may only refer to immutable values
  --> $DIR/const_let_assign3.rs:16:5
   |
LL |     s.foo(3); //~ ERROR references in constants may only refer to immutable values
   |     ^ constants require immutable values

error[E0017]: references in constants may only refer to immutable values
  --> $DIR/const_let_assign3.rs:20:45
   |
LL | type Array = [u32; { let mut x = 2; let y = &mut x; *y = 42; *y}];
   |                                             ^^^^^^ constants require immutable values

error[E0019]: constant contains unimplemented expression type
  --> $DIR/const_let_assign3.rs:20:53
   |
LL | type Array = [u32; { let mut x = 2; let y = &mut x; *y = 42; *y}];
   |                                                     ^^^^^^^

error: aborting due to 4 previous errors

Some errors occurred: E0017, E0019.
For more information about an error, try `rustc --explain E0017`.
